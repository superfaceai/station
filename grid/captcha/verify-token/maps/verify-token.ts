/* eslint-disable jest/no-export */

import { SuperfaceTest } from '@superfaceai/testing';

import { buildSuperfaceTest } from '../../../test-config';

export const verifyTokenTest = (provider: string, validResponse: string) => {
  describe(`captcha/verify-token/${provider}`, () => {
    let superface: SuperfaceTest;

    beforeEach(() => {
      superface = buildSuperfaceTest({
        profile: 'captcha/verify-token',
        provider,
      });
    });

    describe('ValidateCaptchaResponse', () => {
      it('should map error', async () => {
        const result = await superface.run({
          useCase: 'ValidateCaptchaResponse',
          input: {
            response: 'invalid response',
            idempotencyKey: '123e4567-e89b-12d3-a456-426614174000',
          },
        });

        expect(() => result.unwrap()).toThrow();
        expect(result).toMatchSnapshot();
      });

      it('should perform successfully', async () => {
        const result = await superface.run({
          useCase: 'ValidateCaptchaResponse',
          input: {
            response: validResponse,
            idempotencyKey: '123e4567-e89b-12d3-a456-426614174000',
          },
        });

        console.debug(result);

        expect(() => result.unwrap()).not.toThrow();
        expect(result).toMatchSnapshot();
      });
    });
  });
};
