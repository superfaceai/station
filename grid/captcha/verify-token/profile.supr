name = "captcha/verify-token"
version = "1.1.0"

"Validates the CAPTCHA provider response from the website's backend."
usecase ValidateCaptchaResponse idempotent  {
  input {
    "The user response token to be verified, provided by CAPTCHA provider's client-side integration on your site."
    response! string!
  
    "Unique key identifying a repeated request in case of connection issue."
    idempotencyKey string!

    "The user's IP address."
    remoteIp string!

    "The site key you expect to see."
    siteKey string!
  }

  result {
    "Timestamp of the challenge."
    challengeTimeStamp! string!
  
    "Hostname of the site where the CAPTCHA was solved."
    hostname! string!

    "The action name for the request."
    action string!

    "The score for the request (0.0 - 1.0). Returned only in case score based CAPTCHA providers like Google reCAPTCHA v3."
    score number!
  }

  error {
    "List of error codes related to the failed token validation."
    errorCodes! [ErrorCode!]!
  
    "Timestamp of the challenge."
    challengeTimeStamp string!
  
    "Hostname of the site where the CAPTCHA was solved."
    hostname string!
  
    "The rate limit error, if applicable."
    rateLimit number

    "A human-readable explanation specific to this occurrence of the problem."
    detail string
  }

  example SuccessExample {
    input {
      response = 'valid token',
      idempotencyKey = '123e4567-e89b-12d3-a456-426614174000',
    }
    result {
      challengeTimeStamp = '2022-02-28T15:14:30.096Z',
      hostname = 'example.com'
    }
  }

  example ErrorExample {
    input {
      response = 'invalid token',
      idempotencyKey = '123e4567-e89b-12d3-a456-426614174000',
    }
    error {
      errorCodes = ["invalid-input-response"]
      challengeTimeStamp = '2022-02-28T15:14:30.096Z',
      hostname = 'example.com'
    }
  }
}

model ErrorCode enum {
  """
  Missing input secret
  The secret parameter was not passed.
  """
  MissingInputSecret = 'missing-input-secret',
  """
  Invalid input secret
  The secret parameter was invalid or did not exist.
  """
  InvalidInputSecret = 'invalid-input-secret',
  """
  Missing input response
  The response parameter was not passed.
  """
  MissingInputResponse = 'missing-input-response',
  """
  Invalid input response
  The response parameter is invalid or has expired.
  """
  InvalidInputResponse = 'invalid-input-response',
  """
  Invalid widget ID
  The widget ID extracted from the parsed site secret key was invalid or did not exist.
  """
  InvalidWidgetId = 'invalid-widget-id',
  """
  Invalid parsed secret
  The secret extracted from the parsed site secret key was invalid.
  """
  InvalidParsedSecret = 'invalid-parsed-secret',
  """
  Bad request
  The request was rejected because it was malformed.
  """
  BadRequest = 'bad-request',
  """
  Timeout or duplicate
  The response parameter has already been validated before.
  """
  TimeoutOrDuplicate = 'timeout-or-duplicate',
  """
  Internal error
  An internal error happened while validating the response. The request can be retried.
  """
  InternalError = 'internal-error'
  """
  Invalid provider response
  The response received from your CAPTCHA is invalid and could not be processed.
  """
  InvalidProviderResponse = 'invalid-provider-response'
  """
  Invalid or already seen response
  The response parameter has already been checked, or has another issue.
  """
  InvalidOrAlreadySeenResponse = 'invalid-or-already-seen-response'
  """
  Not using dummy passcode
  You have used a testing sitekey but have not used its matching secret.
  """
  NotUsingDummyPasscode = 'not-using-dummy-passcode'
  """
  Sitekey secret mismatch
  The sitekey is not registered with the provided secret.
  """
  SiteKeySecretMissmatch = 'sitekey-secret-mismatch'
}
