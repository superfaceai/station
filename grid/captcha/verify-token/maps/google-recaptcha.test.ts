import { verifyTokenTest } from './verify-token';

verifyTokenTest(
  'google-recaptcha',
  '03AL8dmw9i0g20W8PJJs9Al6ywCa363dUWZQcCRH294Y1LfiYi3fAU3rBRUhMXDVIIUOFaOSvVkrH2BxUh8gIMEp5tlGDHCRNVNZDZ7yBgRTD30JE1JtQyWvX9_wExPCxPz_Cf8YCHYq17ZcqFOzHN6RC0CK9pPum0zVcljbokqLHfrFswDdW64m_krRolgL3KR_XLQBjWSvOLeHGrJkFjVkrqACFJTKH9Sn3nS3h4YcY5xqUFIwHYPvNqCmfPdDDZcGwxXI6Lx6ymYZWu4IIRetjgA_LzG_9cLPkYAT948FrrYlcmXFPMdPLsS6qqsrF3Ce7ge3soGEx4FmJBGwaUCDq6CKLeXo7p0eMbxC7ysvQ9Idny7GnGt7fR3wMgYf-tpIHbScTVfCR4WfLbM-XG1qp2E_AOmwULAKMHWwXT9aaX5rX2d1qhrVb4Fhl_BOfLgAy_poqOBKyT981LuRHV93foZk-BsawIWhqgQNGGZXajY9hDsLQkUcYgV8Fzc8K7HkcTeKdy5UFfpjINVwyl4LQ0DpUL3LXszPYxuKqBh1sXzuZUZVLRIIYfXj8CBWWI5gymRZsn3oiR0SXtajjObTgnakGJY4woL_vJ6ibsgCRL1tf_SPAs19TuPiYSfFNdiIarJGW_F9ghbMVoRPYUcR4ClSGXDqpwwQlcB_eNUmt9xNhySl2rLuH7c5kCoz_dOO8jfW76_Gz2ABbza5GW_KI3dIo_omVDwMAd-VCfhE8ARW_wAVTnutVE6SYPK1wXRF6mpLy_nXd-RVz4vXx4SnKEhf0toniwd1PRMz6CIbjNaztX_qNul99udIX3IhWJbJgZGOuuAvJece8__B_dc_2RDFboBXRfIYbggIfxO6oOY0JlHibpjsmzsAjxLF1jJgSSRxJj9kDndbLhLJPxuavNlOYovLG9UBXiChKYb2FWvf0eFqiskH6N2vEqsFD5OXgvmwAK0PHP_auKVk8UEFZR3Y4Q9XTLL_W4cyi7ZSDjgO_EjMxe85koVrH98G61Zv58owNaXVpTYYRK1y22r-S_K1aa8bGVvlsYV3LToDO50OcMEZyYiUNx5Obx9oCisK3UzUwm7mUeGBS2X9-lT7pswsYmN9VYAIbcKJmrhuVd402_1s02ZNHg0hl_XdBAEMFu8XE6OoT9zeU__p6S59Uf3QkI8qM09NrO9flGxbrnCe06kJ533VZ_woPRn8aDqlD8oav5snjfLt73NrT0IaQ3TWhRFWUqgxNSACwe5iHzG6jKL7DoJOetPK9z3ciUcQQiSIx1avMQVrOGPui0EtYCNqArJuBDZR0d8OL65ygMXp1aygPfX3SLzCpS9XJiLpLtvQssNrCaq_osSQSurnETojzxVYzwvDdoKoIvJ4h3seZsl_EwPBgAMhzL8b5roLSQ-LDUGjn1zW70UhXQObIG5QMLie_n0LiQKRgnCqN3Gj_hzGE234vQGAclY6oKIVeLe4FUFXHn'
);
