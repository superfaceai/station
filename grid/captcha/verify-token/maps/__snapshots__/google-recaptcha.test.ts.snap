// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`captcha/verify-token/google-recaptcha ValidateCaptchaResponse should map error 1`] = `
Err {
  "error": "MappedHTTPError: Expected HTTP error
Properties: {
  \\"errorCodes\\": [
    \\"invalid-input-response\\"
  ],
  \\"detail\\": \\"Challenge verification failed with error code(s) invalid-input-response.\\"
}
AST Path: definitions[0].statements[0].responseHandlers[0].statements[2]
Original Map Location: Line 27, column 7",
}
`;

exports[`captcha/verify-token/google-recaptcha ValidateCaptchaResponse should perform successfully 1`] = `
Ok {
  "value": Object {
    "action": "submit",
    "challengeTimeStamp": "2023-06-15T08:55:11Z",
    "hostname": "localhost",
    "score": 0.9,
  },
}
`;
